torch/testing/_internal/common_device_type.py

```diff
@@ -205,46 +212,83 @@ def _get_precision_override(self, test, dtype):
+                    result = test(self, *args)
+                finally:
+                    self.precision = guard_precision
 
-                dtype_test_name = test_name + dtype_str
-                assert not hasattr(cls, dtype_test_name), "Redefinition of test {0}".format(dtype_test_name)
+                return result
 
-                @wraps(test)
-                def instantiated_test(self, test=test, dtype=dtype):
-                    device_arg = cls.get_primary_device() if not hasattr(test, 'num_required_devices') else cls.get_all_devices()
-                    # Sets precision and runs test
-                    # Note: precision is reset after the test is run
-                    guard_precision = self.precision
-                    try :
-                        self.precision = self._get_precision_override(test, dtype)
-                        result = test(self, device_arg, dtype)
-                    finally:
-                        self.precision = guard_precision
+            # wraps with op decorators
+            if op is not None and op.decorators is not None:
+                for decorator in op.decorators:
+                    instantiated_test = decorator(instantiated_test)
 
-                    return result
+            assert not hasattr(cls, test_name), "Redefinition of test {0}".format(test_name)
+            setattr(cls, test_name, instantiated_test)
 
-                setattr(cls, dtype_test_name, instantiated_test)
+        # Handles tests using the ops decorator
+        if hasattr(test, "op_list"):
+            for op in test.op_list:
+                # Acquires dtypes, using the op data if unspecified
+                dtypes = cls._get_dtypes(test)
+                if dtypes is None:
+                    if cls.device_type == 'cpu' and op.dtypesIfCPU is not None:
+                        dtypes = op.dtypesIfCPU
+                    elif (cls.device_type == 'cuda' and not TEST_WITH_ROCM
+                          and op.dtypesIfCUDA is not None):
+                        dtypes = op.dtypesIfCUDA
+                    elif (cls.device_type == 'cuda' and TEST_WITH_ROCM
+                          and op.dtypesIfROCM is not None):
+                        dtypes = op.dtypesIfROCM
+                    else:
+                        dtypes = op.dtypes
+
+                # Inverts dtypes if the function wants unsupported dtypes
+                if test.unsupported_dtypes_only is True:
+                    dtypes = [d for d in get_all_dtypes() if d not in dtypes]
+
+                dtypes = dtypes if dtypes is not None else (None,)
+                for dtype in dtypes:
+                    instantiate_test_helper(cls,
+                                            name,
+                                            test=test,
+                                            dtype=dtype,
+                                            op=op)
+        else:
+            # Handles tests that don't use the ops decoartor
```

Comment:
This else branch looks like a smell. Below, you treat non-specified dtypes to simply be a request to run the dtype list `(None,)`. It seems to me like ops should be defaulted similarly? Then you skip line 263 when op is None (and then you get the behavior below, without having to replicate it out in this way.)

Reply:
You're correct and this code anticipates a third branch (the `@variants`) branch. It could be optimized better for this case.